intercom-cpp/src/detail/utility.h

```diff
@@ -0,0 +1,77 @@
+
+#ifndef INTERCOM_CPP_DETAIL_UTILITY_H
+#define INTERCOM_CPP_DETAIL_UTILITY_H
+
+
+#include <cstring>
+#include <iomanip>
+#include <sstream>
+#include <type_traits>
+
+namespace intercom
+{
+namespace detail
+{
+
+    /**
+     * @brief Tests whether the machine is little endian or not.
+     *
+     * @return true The machine is little endian.
+     * @return false The machine is big endian.
+     */
+    inline bool is_little_endian()
+    {
+        // Create a copy of the data with memcpy.
+        // Directly casting with e.g. reinterpret_cast will lead to UB.
+        int test_value = 1;
+        std::array< uint8_t, sizeof( int ) > asBytes;
+        std::memcpy( &asBytes, &test_value, sizeof( int ) );
+        return asBytes[ 0 ] == 1;
```

Comment:
`*reinterpret_cast<*char>(..)` should not result in UB. It's one of the few exceptions. Note that `uint8_t` and `char` are different types, so while they have the same size, `uint8_t` reinterpret cast is still UB.

Of course on the other hand, `char` might not be 8-bit type, but at that point `uint8_t` wouldn't exist either as `char` is required to be the smallest type as far as I've understood.

So while using `char` for `is_little_endian` checks, using it for the actual hex-formatting might not be fully cross platform.